lib/recurly/pricing/subscription/tiered-pricing-calculator.js

```diff
@@ -0,0 +1,141 @@
+ * @param {string} currencyCode
+ * @returns {array<object>}
+ *
+ */
+export function getTieredPricingValues (
+  addOn,
+  selectedNumUnits,
+  currencyCode,
+) {
+  const unformattedTiers = addOn.tiers; 
+  return unformattedTiers?.map((tier, index) => {
+    const isFirstTier = index === 0;
+    const isLastTier = index === unformattedTiers.length - 1;
+    const prevTier = isFirstTier ? null : unformattedTiers[index - 1];
+    const tierEndingQuantity = tier.ending_quantity;
+    const endingValue = isLastTier ? null : tierEndingQuantity;
+    const tierCurrency = tier.currencies.find((tc) => tc.currency == currencyCode);
+    const tierValue = tierCurrency.unit_amount;
+    let startingValue;
+
+    // const startingValue = isFirstTier ? 1 : prevTier.ending_quantity + 1;
+    if (isFirstTier) {
+      startingValue = 1;
+    } else {
+      startingValue = prevTier.ending_quantity + 1;
+    }
+    const floatSelectedNumUnits = parseFloat(selectedNumUnits);
+
+    // if floatSelectedNumUnits is less than the startingValue of a tier DONT render it
+    if (floatSelectedNumUnits < startingValue) {
+      return null;
+    }
+
+    let unitAmount = 0;
+    let unitsPerTier = null;
+
+    switch (addOn.tier_type) {
+    case 'tiered':
+      // tier ranges are inclusive. e.g. when tier is 1-10 and floatSelectedNumUnits
+      // is greater than or equal to 10, we want unitsPerTier to be 10,
+      // so max - min + 1 (10 - 1 + 1 === 10)
+      unitsPerTier = floatSelectedNumUnits >= tierEndingQuantity
+        ? tierEndingQuantity - startingValue + 1
+        : floatSelectedNumUnits - startingValue + 1;
+
+      unitAmount = calculateTotal(unitsPerTier, tierValue);
+      break;
+    case 'volume':
+      if (floatSelectedNumUnits <= tierEndingQuantity) {
+        unitAmount = calculateTotal(floatSelectedNumUnits, tierValue);
+      }
+      break;
+    case 'stairstep':
+      if (selectedNumUnits && selectedNumUnits <= tierEndingQuantity) {
+        unitAmount = tierValue;
+      }
+      break;
+    default:
+      break;
+    }
+
+    return {
+      startingValue,
+      endingValue,
+      unitAmount,
+      currencyCode,
+      unitsPerTier,
+    };
+  });
+}
+
+/**
+ *
+ * returns the total estimated cost (unitAmount)
+ * of a `tiered` | `volume` | `stairstep` pricing model
+ *
+ * @param {array} unformattedTiers [{ currency_code, ending_quantity, unit_amount}]
+ * @param {int} selectedNumUnits
+ * @param {string} currencyCode
+ * @returns {object}
+ *
+ */
+export function getTieredPricingTotal (
+  addOn,
+  selectedNumUnits,
+  currencyCode,
+) {
+  const initialTieredPricingTotal = {
+    unitAmount: 0,
+  };
+
+  return getTieredPricingValues(
+    addOn,
+    selectedNumUnits,
+    currencyCode
+  )?.reduce(
+    (tieredPricingTotal, tier) => {
+      if (tier) {
+        tieredPricingTotal.unitAmount += tier.unitAmount;
+      }
+      return tieredPricingTotal;
+    },
+    initialTieredPricingTotal,
+  );
+}
```

Comment:
I think it would be beneficial to compose unit tests for this module specifically, since the e2e-leaning pricing coverage below may not test some of the routines in here.

Reply:
Would it be better to write the test for the module or just copy the newest tiered tests to cover volume and stairstep?

Reply:
🤔 definitely good to do both. But the priority would be unit testing of this module.